### ISSUE TITLE
Bot "Send email to recipient" behaviour changed in 9.5.

### ISSUE THREAD

**Issue by iammattmartin** [jstanden/cerb#1291]
Prior to 9.5 when using this function and ensuring "Don't save a copy of this message in conversation history", the message would be sent and the ticket status would not be changed.

Now it is changing the ticket back to open which shouldn't happen.

In our use case, we reject some direct to email messages when no existing ticket exists. We send an advisory message on the ticket already deleted, so then that being re-opened is not expected or ideal.

Message log.

> Auto-Reply Bot [Bounce and Delete (Sales)] changed ticket to status open
> 4 secs ago
> Auto-Reply Bot [Bounce and Delete (Sales)] responded to ticket
> 5 secs ago
> Auto-Reply Bot [Drop / Store Source / Auto-Reply] changed ticket to status deleted
> 5 secs ago
> Cerb moved ticket to Sales: Inbox
> 6 secs ago
> Me <me@---.com> wrote on ticket
> 6 secs ago

Before it would not get to the re-open step.

**Comment by jstanden** [comment]
Can you export those behaviors and send them to team@cerb.ai ?

I'm attempting to reproduce the issue, but can't get the 'Send email to recipients' action to reopen the ticket.

**Comment by iammattmartin** [comment]
Sure. It is pretty hefty though. On its way.

**Comment by jstanden** [comment]
Thanks!

I'll pare it down to just the path we're testing here. Knowing the second behavior is scheduled helps.

**Comment by iammattmartin** [comment]
Think you should be able to just run the smaller scheduled task that appears to be what calls it, at least judging from the logs and my further testing.

**Comment by jstanden** [comment]
Yeah, I can reproduce it by just scheduling the second behavior with 'Send email to recipients'.

If I delete the ticket beforehand, then run the scheduler, it reopens the ticket.

**Comment by iammattmartin** [comment]
At least i'm not going crazy then. ;)

**Comment by iammattmartin** [comment]
Obviously, for "Send email to recipients" when adding that message to the ticket, I would expect it to change back to open or waiting but not when you're leaving the ticket alone without further updates.

Or maybe as a bot, it shouldn't even do that.... (change status in either situation) unless you have a rule that sets status depending on a response in or out.

**Comment by iammattmartin** [comment]
@jstanden Can you see anyway to easily work around this so we can upgrade for the security benefits from 9.4, apart from disabling all bots that use the function....

**Comment by jstanden** [comment]
@iammattmartin We should probably just add a 'Status:' option to that action to make the outcome easily controllable.

**Comment by iammattmartin** [comment]
Well, the "old" way it worked, just worked well. Is this new behaviour expected or just due to some other inner workings changing?

**Comment by jstanden** [comment]
I'll verify what's going on when scheduled first.

**Comment by jstanden** [comment]
k! Apologies for the delay here. I definitively found the issue. I could repro in one dev environment and not another clean one.

When there are any 'Before sent' behaviors, they initialize `null` default properties on the pending draft (e.g. status, to, cc, reopen_at) if not already set via references. This may be a side effect on specific PHP versions.

There's new code on tickets handling message events, which now checks those properties for the _existence_ of the `status_id` key then typecasts to an int. That means the null turns into `0`, which equates to "open".

I added a quick stopgap to check if that property was null and properly ignore it. The better long -term fix is making sure 'Before sent' events don't unexpectedly initialize null values for unset properties.

Kludgy stopgap:

~~~
Index: features/cerberusweb.core/api/dao/ticket.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- features/cerberusweb.core/api/dao/ticket.php	(revision d063d473ddc4c79a0293429f63f49fbe539b3d08)
+++ features/cerberusweb.core/api/dao/ticket.php	(date 1589337650996)
@@ -965,8 +965,13 @@
 		}
 		
 		if(array_key_exists('status_id', $properties)) {
-			@$status_id = DevblocksPlatform::importVar($properties['status_id'], 'int', 0);
-			@$reopen_at = DevblocksPlatform::importVar($properties['ticket_reopen'], 'string', '');
+			if(is_null($properties['status_id'])) {
+				$status_id = null;
+				$reopen_at = null;
+			} else {
+				@$status_id = DevblocksPlatform::importVar($properties['status_id'], 'int', 0);
+				@$reopen_at = DevblocksPlatform::importVar($properties['ticket_reopen'], 'string', '');
+			}
 			
 			// Handle reopen date
 			if($reopen_at) {
~~~

Proper fix forthcoming.

**Comment by iammattmartin** [comment]
I applied this but it didn't seem to work. A ticket that was in waiting status was set back to open when it was replied to (not stored) by a bot for the 7 days closure warning.

Not sure if this will only apply to tickets replied to after this was applied?

**Comment by jstanden** [comment]
This should be properly fixed in 9.5.7 now.

~~~
commit c43623989fdc6a93e0be94c9b51faf501ccb4f7a
Author: Jeff Standen <jeff@webgroupmedia.com>
Date:   Thu Jun 4 19:05:52 2020 -0700

    * [Mail/Behaviors] Fixed a regression in 'Before message sent by worker' behaviors. These could unexpectedly re-open tickets on scheduled replies.

diff --git a/features/cerberusweb.core/api/dao/ticket.php b/features/cerberusweb.core/api/dao/ticket.php
index ea9338c58..0c5c231ac 100644
--- a/features/cerberusweb.core/api/dao/ticket.php
+++ b/features/cerberusweb.core/api/dao/ticket.php
@@ -964,14 +964,9 @@ class DAO_Ticket extends Cerb_ORMHelper {
 				unset($properties['owner_id']);
 		}
 		
-		if(array_key_exists('status_id', $properties)) {
-			if(is_null($properties['status_id'])) {
-				$status_id = null;
-				$reopen_at = null;
-			} else {
-				@$status_id = DevblocksPlatform::importVar($properties['status_id'], 'int', 0);
-				@$reopen_at = DevblocksPlatform::importVar($properties['ticket_reopen'], 'string', '');
-			}
+		if(array_key_exists('status_id', $properties) && !is_null($properties['status_id'])) {
+			@$status_id = DevblocksPlatform::importVar($properties['status_id'], 'int', 0);
+			@$reopen_at = DevblocksPlatform::importVar($properties['ticket_reopen'], 'string', '');
 			
 			// Handle reopen date
 			if($reopen_at) {

~~~
